File: programming_assignments/beale_ciphers/solution/solve.py
import time
import random


def freivald(X, Y, Q, x, y, n, m, k, v, t, q):
    count = 0
    while True:
        v = [random.randint(0, 1) for i in range(k)]
        for i in range(k):
            t[i] = 0
            q[i] = 0
            for j in range(k):
                t[i] = t[i] + Q[x + i][y + j] * v[j]
                q[i] = q[i] + Y[i][j] * v[j]
        for i in range(k):
            v[i] = 0
            for j in range(k):
                v[i] = v[i] + X[i][j] * t[j]

        for i in range(k):
            if v[i] != q[i]:
                return False
        if count > 10:
            break
        count = count + 1
    return True


def solve():
    n, m, k = map(int, input().split())
    X = []
    Y = []
    Q = []
    v = [0 for i in range(k)]
    t = [0 for i in range(k)]
    q = [0 for i in range(k)]

    for i in range(k):
        X.append(map(int, input().split()))
    for i in range(k):
        Y.append(map(int, input().split()))
    for i in range(n):
        Q.append(map(int, input().split()))

    ans = 0
    for i in range(n - k + 1):
        for j in range(m - k + 1):
            if freivald(X, Y, Q, i, j, n, m, k, v, t, q):
                ans = ans + 1
    print(ans)


def test():
    T = map(int, input().split())[0]
    for i in range(T):
        solve()


if __name__ == '__main__':
    test()

File: programming_assignments/beale_ciphers/solution/solve.java
import java.util.*;

public class Main{
    public static void main(String args[]){
        Task solver = new Task();
        solver.solve();
    }
}

class Task{
    public void solve(){
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();

        for(int R=1; R<=T; R++)
            Do(sc);
    }

    public boolean freivald(int X[][], int Y[][], int Q[][], int x, int y, int n, int m, int k){
        int T = 10;
        int v[] = new int[k];
        int t[] = new int[k];
        int q[] = new int[k];
        Random rd = new Random();

        while( T-- > 0 ) {
            for(int i=0; i<k; i++)
                v[i] = rd.nextInt(2);

            for(int i=0; i<k; i++){
                t[i] = q[i] = 0;
                for(int j=0; j<k; j++){
                    t[i] += Q[x+i][y+j] * v[j];
                    q[i] += Y[i][j] * v[j];
                }
            }
            for(int i=0; i<k; i++){
                v[i] = 0;
                for(int j=0; j<k; j++)
                    v[i] += X[i][j] * t[j];
            }

            for(int i=0; i<k; i++)
                if( v[i] != q[i] ) return false;

        }
        return true;
    }

    public void Do(Scanner sc){
        int n, m, k;
        n = sc.nextInt();
        m = sc.nextInt();
        k = sc.nextInt();

        int X[][] = new int[k][k];
        int Y[][] = new int[k][k];
        int Q[][] = new int[n][m];

        for(int i=0; i<k; i++)
            for(int j=0; j<k; j++)
                X[i][j] = sc.nextInt();

        for(int i=0; i<k; i++)
            for(int j=0; j<k; j++)
                Y[i][j] = sc.nextInt();

        for(int i=0; i<n; i++)
            for(int j=0; j<m; j++)
                Q[i][j] = sc.nextInt();

        int ans = 0;
        for(int i=0; i<n-k+1; i++)
            for(int j=0; j<m-k+1; j++)
                if(freivald( X, Y, Q, i, j, n, m, k ))
                    ++ans;

        System.out.println(ans);
    }
}
